File: src/main/java/ru/org/linux/user/UserLogDao.java
/*
 * Copyright 1998-2022 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user;

import com.google.common.collect.ImmutableMap;
import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import javax.annotation.Nonnull;
import javax.sql.DataSource;
import java.time.Duration;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.sql.Timestamp;
import java.util.Optional;

@Repository
public class UserLogDao {
  public static final String OPTION_OLD_USERPIC = "old_userpic";
  public static final String OPTION_NEW_USERPIC = "new_userpic";
  public static final String OPTION_BONUS = "bonus";
  public static final String OPTION_REASON = "reason";
  public static final String OPTION_OLD_EMAIL = "old_email";
  public static final String OPTION_NEW_EMAIL = "new_email";
  public static final String OPTION_OLD_INFO = "old_info";
  public static final String OPTION_OLD_TOWN = "old_town";
  public static final String OPTION_OLD_URL = "old_url";
  public static final String OPTION_IP = "ip";
  public static final String OPTION_USET_AGENT = "user_agent";
  public static final String OPTION_INVITED_BY = "invited_by";

  private JdbcTemplate jdbcTemplate;

  @Autowired
  private void setDataSource(DataSource ds) {
    jdbcTemplate = new JdbcTemplate(ds);
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logResetUserpic(@Nonnull User user, @Nonnull User actionUser, int bonus) {
    ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();

    if (bonus!=0) {
      builder.put(OPTION_BONUS, bonus);
    }

    if (user.getPhoto()!=null) {
      builder.put(OPTION_OLD_USERPIC, user.getPhoto());
    }

    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            actionUser.getId(),
            UserLogAction.RESET_USERPIC.toString(),
            builder.build()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logSetUserpic(@Nonnull User user, @Nonnull String userpic) {
    ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();

    if (user.getPhoto()!=null) {
      builder.put(OPTION_OLD_USERPIC, user.getPhoto());
    }

    builder.put(OPTION_NEW_USERPIC, userpic);

    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            user.getId(),
            UserLogAction.SET_USERPIC.toString(),
            builder.build()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logBlockUser(@Nonnull User user, @Nonnull User moderator, @Nonnull String reason) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.BLOCK_USER.toString(),
            ImmutableMap.of(OPTION_REASON, reason)
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logFreezeUser(@Nonnull User user, @Nonnull User moderator,  
    @Nonnull String reason, @Nonnull Timestamp until) {

    Timestamp     now = new Timestamp(System.currentTimeMillis());
    UserLogAction action = UserLogAction.FROZEN;

    // the action may be not consistent with database (e.g. with real action)
    // if the 'until' is close to the now, but we don't have to worry about it,
    // since, it's not about real use cases
    if (until.before(now)) {
        action = UserLogAction.DEFROSTED;
    }

    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            action.toString(),
            ImmutableMap.of(OPTION_REASON, reason)
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logScore50(@Nonnull User user, @Nonnull User moderator) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, '')",
            user.getId(),
            moderator.getId(),
            UserLogAction.SCORE50.toString()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logUnblockUser(@Nonnull User user, @Nonnull User moderator) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.UNBLOCK_USER.toString(),
            ImmutableMap.of()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logAcceptNewEmail(@Nonnull User user, @Nonnull String newEmail) {
    ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();

    builder.put(OPTION_NEW_EMAIL, newEmail);

    if (user.getEmail()!=null) {
      builder.put(OPTION_OLD_EMAIL, user.getEmail());
    }

    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            user.getId(),
            UserLogAction.ACCEPT_NEW_EMAIL.toString(),
            builder.build()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logResetInfo(@Nonnull User user, @Nonnull User moderator, @Nonnull String userInfo, int bonus) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.RESET_INFO.toString(),
            ImmutableMap.of(
                    OPTION_OLD_INFO, userInfo,
                    OPTION_BONUS, bonus
            )
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logResetUrl(@Nonnull User user, @Nonnull User moderator, @Nonnull String url, int bonus) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.RESET_URL.toString(),
            ImmutableMap.of(
                    OPTION_OLD_URL, url,
                    OPTION_BONUS, bonus
            )
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logResetTown(@Nonnull User user, @Nonnull User moderator, @Nonnull String town, int bonus) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.RESET_TOWN.toString(),
            ImmutableMap.of(
                    OPTION_OLD_TOWN, town,
                    OPTION_BONUS, bonus
            )
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logResetPassword(@Nonnull User user, @Nonnull User moderator) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.RESET_PASSWORD.toString(),
            ImmutableMap.of()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logSetPassword(@Nonnull User user) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            user.getId(),
            UserLogAction.SET_PASSWORD.toString(),
            ImmutableMap.of()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logSetUserInfo(@Nonnull User user, Map<String, String> info) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            user.getId(),
            UserLogAction.SET_INFO.toString(),
            info
    );
  }

  @Nonnull
  public List<UserLogItem> getLogItems(@Nonnull User user, boolean includeSelf) {
    String sql =
            includeSelf ?
            "SELECT id, userid, action_userid, action_date, action, info FROM user_log WHERE userid=? ORDER BY id DESC"
            :
            "SELECT id, userid, action_userid, action_date, action, info FROM user_log WHERE userid=? AND userid!=action_userid ORDER BY id DESC";

    return jdbcTemplate.query(
            sql,
            (rs, rowNum) -> new UserLogItem(
                    rs.getInt("id"),
                    rs.getInt("userid"),
                    rs.getInt("action_userid"),
                    new DateTime(rs.getTimestamp("action_date")),
                    UserLogAction.valueOf(rs.getString("action").toUpperCase()),
                    (Map<String, String>) rs.getObject("info")
            ),
            user.getId()
    );
  }

  public int getUserpicSetCount(User user, Duration duration) {
    return jdbcTemplate.queryForObject(
            "SELECT count(*) FROM user_log WHERE userid=? AND action=?::user_log_action AND action_date>?",
            Integer.class,
            user.getId(),
            UserLogAction.SET_USERPIC.toString(),
            OffsetDateTime.now().minus(duration));
  }

  public boolean hasRecentModerationEvent(User user, Duration duration, UserLogAction action) {
    return jdbcTemplate.queryForObject(
            "SELECT EXISTS (SELECT * FROM user_log WHERE userid=? AND action=?::user_log_action AND action_date>? AND userid!=action_userid)",
            Boolean.class,
            user.getId(),
            action.toString(),
            OffsetDateTime.now().minus(duration));
  }

  public List<Integer> getRecentlyHasEvent(UserLogAction action) {
    return jdbcTemplate.queryForList(
            "SELECT userid FROM user_log WHERE action=?::user_log_action AND action_date>CURRENT_TIMESTAMP - interval '3 days' ORDER BY action_date",
            Integer.class,
            action.toString());
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void logRegister(int userid, @Nonnull String ip, Optional<Integer> invitedBy, int userAgent) {
    ImmutableMap<String, String> params;

    params = invitedBy
            .map(integer -> ImmutableMap.of(OPTION_IP, ip, OPTION_USET_AGENT, Integer.toString(userAgent), OPTION_INVITED_BY, integer.toString()))
            .orElseGet(() -> ImmutableMap.of(OPTION_IP, ip, OPTION_USET_AGENT, Integer.toString(userAgent)));

    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            userid,
            userid,
            UserLogAction.REGISTER.toString(),
            params
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void setCorrector(@Nonnull User user, @Nonnull User moderator) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.SET_CORRECTOR.toString(),
            ImmutableMap.of()
    );
  }

  @Transactional(rollbackFor = Exception.class, propagation = Propagation.MANDATORY)
  public void unsetCorrector(@Nonnull User user, @Nonnull User moderator) {
    jdbcTemplate.update(
            "INSERT INTO user_log (userid, action_userid, action_date, action, info) VALUES (?,?,CURRENT_TIMESTAMP, ?::user_log_action, ?)",
            user.getId(),
            moderator.getId(),
            UserLogAction.UNSET_CORRECTOR.toString(),
            ImmutableMap.of()
    );
  }

}

File: src/main/scala/ru/org/linux/user/UserLogDao.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user

import org.springframework.stereotype.Repository
import ru.org.linux.scalikejdbc.{SpringDB, Transaction}
import ru.org.linux.scalikejdbc.SpringDB.given
import ru.org.linux.scalikejdbc.Transaction.given
import scalikejdbc.*

import java.time.{Duration, Instant, OffsetDateTime}
import java.util as ju
import javax.annotation.Nullable
import scala.jdk.CollectionConverters.*

@Repository
class UserLogDao(springDB: SpringDB):

  private def insertLog(userid: Int, actionUserId: Int, action: UserLogAction, info: ju.Map[String, String])(using
      DBSession): Unit =
    sql"""INSERT INTO user_log (userid, action_userid, action_date, action, info)
          VALUES ($userid, $actionUserId, CURRENT_TIMESTAMP, ${action.toDbName}::user_log_action, $info)"""
      .update
      .apply()

  def logResetUserpic(user: User, actionUser: User, bonus: Int)(using Transaction): Unit =
    val info: ju.Map[String, String] = (
      Map.empty[String, String] ++ Option.when(bonus != 0)(UserLogDao.OptionBonus -> bonus.toString) ++
        Option.when(user.photo != null)(UserLogDao.OptionOldUserpic -> user.photo)
    ).asJava

    insertLog(user.id, actionUser.id, UserLogAction.ResetUserpic, info)

  def logSetUserpic(user: User, userpic: String)(using Transaction): Unit =
    val info: ju.Map[String, String] = (
      Map.empty[String, String] ++ Option.when(user.photo != null)(UserLogDao.OptionOldUserpic -> user.photo) +
        (UserLogDao.OptionNewUserpic -> userpic)
    ).asJava

    insertLog(user.id, user.id, UserLogAction.SetUserpic, info)

  def logBlockUser(user: User, moderator: User, reason: String)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.BlockUser, Map(UserLogDao.OptionReason -> reason).asJava)

  def logFreezeUser(user: User, moderator: User, reason: String, until: Instant)(using Transaction): Unit =
    val action =
      if until.isBefore(Instant.now()) then
        UserLogAction.Defrosted
      else
        UserLogAction.Frozen

    val info: ju.Map[String, String] =
      if until.isBefore(Instant.now()) then
        Map(UserLogDao.OptionReason -> reason).asJava
      else
        Map(UserLogDao.OptionReason -> reason, UserLogDao.OptionUntil -> until.toString).asJava

    insertLog(user.id, moderator.id, action, info)

  def logScore50(user: User, moderator: User)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.Score50, ju.Collections.emptyMap[String, String]())

  def logUnblockUser(user: User, moderator: User)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.UnblockUser, ju.Collections.emptyMap[String, String]())

  def logAcceptNewEmail(user: User, newEmail: String)(using Transaction): Unit =
    val info: ju.Map[String, String] = (
      Map(UserLogDao.OptionNewEmail -> newEmail) ++
        Option.when(user.email != null)(UserLogDao.OptionOldEmail -> user.email)
    ).asJava

    insertLog(user.id, user.id, UserLogAction.AcceptNewEmail, info)

  def logResetInfo(user: User, moderator: User, userInfo: String, bonus: Int)(using Transaction): Unit =
    val info = Map(UserLogDao.OptionOldInfo -> userInfo, UserLogDao.OptionBonus -> bonus.toString).asJava
    insertLog(user.id, moderator.id, UserLogAction.ResetInfo, info)

  def logResetUrl(user: User, moderator: User, url: String, bonus: Int)(using Transaction): Unit =
    val info = Map(UserLogDao.OptionOldUrl -> url, UserLogDao.OptionBonus -> bonus.toString).asJava
    insertLog(user.id, moderator.id, UserLogAction.ResetUrl, info)

  def logResetTown(user: User, moderator: User, town: String, bonus: Int)(using Transaction): Unit =
    val info = Map(UserLogDao.OptionOldTown -> town, UserLogDao.OptionBonus -> bonus.toString).asJava
    insertLog(user.id, moderator.id, UserLogAction.ResetTown, info)

  def logSentPasswordReset(
      resetFor: User,
      @Nullable
      resetBy: User,
      email: String)(using Transaction): Unit =
    val actionUserId =
      if resetBy != null then
        resetBy.id
      else
        resetFor.id
    insertLog(resetFor.id, actionUserId, UserLogAction.SentPasswordReset, Map(UserLogDao.OptionEmail -> email).asJava)

  def logResetPassword(user: User, moderator: User)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.ResetPassword, ju.Collections.emptyMap[String, String]())

  def logSetPassword(user: User, ip: String)(using Transaction): Unit =
    insertLog(user.id, user.id, UserLogAction.SetPassword, Map(UserLogDao.OptionIp -> ip).asJava)

  def logSetUserInfo(user: User, info: ju.Map[String, String])(using Transaction): Unit =
    insertLog(user.id, user.id, UserLogAction.SetInfo, info)

  def getLogItems(user: User, includeSelf: Boolean): Seq[UserLogItem] =
    springDB.run {
      if includeSelf then
        sql"""SELECT id, userid, action_userid, action_date, action, info
            FROM user_log WHERE userid=${user.id} ORDER BY id DESC""".map(toLogItem).list.apply()
      else
        sql"""SELECT id, userid, action_userid, action_date, action, info
            FROM user_log WHERE userid=${user.id} AND userid!=action_userid ORDER BY id DESC"""
          .map(toLogItem)
          .list
          .apply()
    }

  def getUserpicSetCount(user: User, duration: Duration): Int =
    springDB.run {
      sql"""SELECT count(*) FROM user_log
          WHERE userid=${user.id} AND action=${UserLogAction
          .SetUserpic
          .toDbName}::user_log_action AND action_date>${OffsetDateTime.now().minus(duration)}"""
        .map(rs => rs.int(1))
        .single
        .apply()
        .getOrElse(0)
    }

  def hasRecentModerationEvent(user: User, duration: Duration, action: UserLogAction): Boolean =
    springDB.run {
      sql"""SELECT EXISTS (SELECT * FROM user_log
          WHERE userid=${user.id} AND action=${action.toDbName}::user_log_action AND action_date>${OffsetDateTime
          .now()
          .minus(duration)} AND userid!=action_userid)""".map(rs => rs.boolean(1)).single.apply().getOrElse(false)
    }

  def hasRecentSelfEvent(user: User, duration: Duration, action: UserLogAction): Boolean =
    springDB.run {
      sql"""SELECT EXISTS (SELECT * FROM user_log
          WHERE userid=${user.id} AND action=${action.toDbName}::user_log_action AND action_date>${OffsetDateTime
          .now()
          .minus(duration)} AND userid=action_userid)""".map(rs => rs.boolean(1)).single.apply().getOrElse(false)
    }

  def getRecentlyHasEvent(action: UserLogAction): Seq[Int] =
    springDB.run {
      sql"""SELECT userid FROM user_log
          WHERE action=${action
          .toDbName}::user_log_action AND action_date>CURRENT_TIMESTAMP - interval '3 days' ORDER BY action_date"""
        .map(rs => rs.int("userid"))
        .list
        .apply()
    }

  def logRegister(userid: Int, ip: String, userAgent: Int, language: Option[String])(using Transaction): Unit =
    val info: ju.Map[String, String] = (
      Map(UserLogDao.OptionIp -> ip, UserLogDao.OptionUserAgent -> userAgent.toString) ++
        language.map(lang => UserLogDao.OptionAcceptLanguage -> lang)
    ).asJava

    insertLog(userid, userid, UserLogAction.Register, info)

  def setCorrector(user: User, moderator: User)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.SetCorrector, ju.Collections.emptyMap[String, String]())

  def unsetCorrector(user: User, moderator: User)(using Transaction): Unit =
    insertLog(user.id, moderator.id, UserLogAction.UnsetCorrector, ju.Collections.emptyMap[String, String]())

  private def toLogItem(rs: WrappedResultSet): UserLogItem =
    val options =
      rs.get[ju.Map[String, String]]("info")
        .asScala
        .view
        .mapValues(v =>
          if v == null then
            ""
          else
            v)
        .toMap
    UserLogItem(
      rs.int("id"),
      rs.int("userid"),
      rs.int("action_userid"),
      rs.timestamp("action_date").toInstant,
      UserLogAction.fromDbName(rs.string("action")),
      options)

object UserLogDao:
  val OptionOldUserpic = "old_userpic"
  val OptionNewUserpic = "new_userpic"
  val OptionBonus = "bonus"
  val OptionReason = "reason"
  val OptionEmail = "email"
  val OptionOldEmail = "old_email"
  val OptionNewEmail = "new_email"
  val OptionOldInfo = "old_info"
  val OptionOldTown = "old_town"
  val OptionOldUrl = "old_url"
  val OptionIp = "ip"
  val OptionUserAgent = "user_agent"
  val OptionInvitedBy = "invited_by"
  val OptionAcceptLanguage = "accept_lang"
  val OptionUntil = "until"
